### PR TITLE
[SPARK-14217][SQL] Fix bug if parquet data has columns that use dictionary encoding for some of the data

### PR DESCRIPTION
**Pull request by nongli**
## What changes were proposed in this pull request?

Currently, this causes batches where some values are dictionary encoded and some
which are not. The non-dictionary encoded values cause us to remove the dictionary
from the batch causing the first values to return garbage.

This patch fixes the issue by first decoding the dictionary for the values that are
already dictionary encoded before switching. A similar thing is done for the reverse
case where the initial values are not dictionary encoded.
## How was this patch tested?

This is difficult to test but replicated on a test cluster using a large tpcds data set.

### CODE REVIEW

**Comment by rxin** [comment]
cc @sameeragarwal

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 1]
```diff
@@ -246,11 +215,48 @@ private void decodeDictionaryIds(int rowId, int num, ColumnVector column,
...
+          for (int i = rowId; i < rowId + num; ++i) {
+            column.putByte(i, (byte) dictionary.decodeToInt(dictionaryIds.getInt(i)));
+          }
+        } else if (column.dataType() == DataTypes.ShortType) {
+          for (int i = rowId; i < rowId + num; ++i) {
+            column.putShort(i, (short) dictionary.decodeToInt(dictionaryIds.getInt(i)));
+          }
+        } else if (DecimalType.is64BitDecimalType(column.dataType())) {
+          for (int i = rowId; i < rowId + num; ++i) {
+            column.putLong(i, dictionary.decodeToInt(dictionaryIds.getInt(i)));
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: we use int for small decimal

**Comment by davies** [comment]
@nongli Is this because that within the same group, some pages are dictionary encoded, some not?

**Comment by nongli** [comment]
@davies Yea. The format allows at most 1 dictionary per column per row group. Each page can have a different encoding though.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 2]
```diff
@@ -246,11 +215,48 @@ private void decodeDictionaryIds(int rowId, int num, ColumnVector column,
...
+          }
+        } else if (column.dataType() == DataTypes.ByteType) {
+          for (int i = rowId; i < rowId + num; ++i) {
+            column.putByte(i, (byte) dictionary.decodeToInt(dictionaryIds.getInt(i)));
+          }
+        } else if (column.dataType() == DataTypes.ShortType) {
+          for (int i = rowId; i < rowId + num; ++i) {
+            column.putShort(i, (short) dictionary.decodeToInt(dictionaryIds.getInt(i)));
+          }
+        } else if (DecimalType.is64BitDecimalType(column.dataType())) {
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: is32BitDecimalType ?

**Comment by yhuai** [comment]
Is this PR good to go? Or we still need to update it?

**Comment by yhuai** [comment]
test this please

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 3]
```diff
@@ -200,9 +150,28 @@ void readBatch(int total, ColumnVector column) throws IOException {
...
+            descriptor.getType() == PrimitiveType.PrimitiveTypeName.BINARY)) {
+          // Column vector supports lazy decoding of dictionary values so just set the dictionary.
+          // We can't do this if rowId != 0 AND the column doesn't have a dictionary (i.e. some
+          // non-dictionary encoded values have already been added).
+          column.setDictionary(dictionary);
+        } else {
+          decodeDictionaryIds(rowId, num, column, dictionaryIds);
+        }
       } else {
-        column.setDictionary(null);
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: I think we should always have this, if column.hasDictionary(), or getLong() will use the dictionary inside the column.

**Comment by davies** [comment]
@nongli Do you need me to take over this patch?

**Comment by yhuai** [comment]
@davies Can you help us update this patch? Thanks!

**Comment by srowen** [comment]
@nongli I think you can close this in favor of https://github.com/apache/spark/pull/12279

**Pull request closed without merging**